ms_active_directory/core/ad_session.py

```diff
@@ -1686,10 +1689,18 @@ def find_groups_for_entities(self, entities: List[Union[str, ADObject]], attribu
                                                                              skip_validation)
 
         filter_pieces = []
-        for entity_dn in entity_dns_to_entities:
+        mapping_dict = {}
+        for entity_dn, entity in entity_dns_to_entities:
             filter_piece = '({}={})'.format(ldap_constants.AD_ATTRIBUTE_MEMBER,
                                             ldap_utils.escape_dn_for_filter(entity_dn))
             filter_pieces.append(filter_piece)
+            mapping_dict[entity] = []
+            if include_primary:
+                entity_obj = self.find_object_by_distinguished_name(entity_dn, controls=controls)
+                # only get primary group if entity is of type ADUser or ADComputer
+                if isinstance(entity_obj, ADUser) or isinstance(entity_obj, ADComputer):
+                    primary_group = self.find_primary_group_for_account(entity_obj, attributes_to_lookup, controls)
+                    mapping_dict[entity].append(primary_group)
```

Comment:
this could be optimized a good amount. we already potentially did these lookups in `normalize_entities_to_entity_dns` so doing them again, and one at a time, is pretty costly and removes the performance benefit of skipping validation.

even ignoring that - we do 2 lookups here and it seems one is basically just "does this thing exist and is it a user or computer", and the second is "go get attributes about the object" and then "go find the primary group".

I think it would make a lot of sense to grab the SID and primary group id in our first lookup here (it's fine if it's not a user/computer because we won't error trying to read non-existent fields).
then we can create some internal `_find_primary_group_for_sid_and_group_id` that assumes we already have the sid and rid and go from there.
building on that, it would also potentially be helpful to include some caching and some checks for if the primary group's SID is one of our `WellKnownSID`s (e.g. `Domain Administrators`, `Account Operators`, etc.) because then we can significantly improve our performance and reduce the number of lookups.
the primary group is just going to be `Domain Users` by default in the majority of scenarios (since primary group was added for some special cases like network services and posix compatibility) so caching the result for looking up the primary group within a single run of this function could significantly improve performance

does that seem reasonable?

Reply:
I think I may have come up with a better way to handle this. Instead of actually looking up each group and caching it, I instead add a filter for the `objectSid` to the chain filter so it can find all the groups in a single query. I also changed `filter_pieces` from a list to a set to prevent duplicate filters. 

Let me know what you think of the changes in my latest commits.